### PR TITLE
[TR] PIM-4227: Disable product versionning on category update

### PR DESCRIPTION
**Pull request by phaseinducer**
| Q                    | A
| -------------------- | ---
| Bug fix?             | Y
| New feature?         | N
| BC breaks?           | N
| CI currently passes? | Y
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | Y
| Fixed tickets        | PIM-4227
| Doc PR               |

At category update : don't create new products versions anymore.
At category removal : check if the category and all sub-categories contain more products than a limit specified in parameters, if so, display an alert and don't remove the category.

### CODE REVIEW

**Comment by jjanvier** [comment]
ideally we should have a behat that there is no new product version when a category is updated

**Comment by phaseinducer** [comment]
@jjanvier : I'm on it for the removal, but for the update I don't think it's possible. It uses drag'n'drop and there something in Mink and / or Selenium driver that mess up. The scenario exists but is currently skipped : https://github.com/akeneo/pim-community-dev/blob/1.2/features/category/drag_and_drop_a_category.feature

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 1]
```diff
@@ -21,13 +21,19 @@ class CategoryExtension extends \Twig_Extension
     protected $manager;
 
     /**
+     * @var integer
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: we declare 'var' on a single line as /** @var integer */

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 2]
```diff
@@ -21,13 +21,19 @@ class CategoryExtension extends \Twig_Extension
...
+     */
+    protected $productsLimitForRemoval;
+
+    /**
      * Constructor
      *
      * @param ProductCategoryManager $manager
      */
-    public function __construct(ProductCategoryManager $manager)
+    public function __construct(ProductCategoryManager $manager, $productsLimitForRemoval)
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: add a default value for $productsLimitForRemoval a partner has override the service (and drop it when merge on master)

**Comment by sumbobyboys** [comment]
GTM

File: src/Pim/Bundle/EnrichBundle/Resources/views/CategoryTree/form.html.twig [hunk 3]
```diff
@@ -27,15 +27,31 @@
...
-                ~ 'info.category.keep products'|trans,
-                (form.vars.value.parent ? 'flash.category.removed' : 'flash.tree.removed')|trans
-            ) }}
+            {% if category_exceeds_products_limit_for_removal(form.vars.value, true) %}
+                {{ elements.link(
+                    'btn.delete',
+                    null,
+                    {
+                        'icon':           'trash',
+                        'title':          'btn.delete',
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Resources/views/CategoryTree/form.html.twig]: remove extra spaces between keys and values

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 4]
```diff
@@ -162,6 +170,29 @@ public function listCategoriesResponse(
...
+     * Check if specified category exceeds the products limit for removal
+     *
+     * @param CategoryInterface $category
+     * @param boolean           $includeSub
+     *
+     * @return boolean
+     */
+    public function categoryExceedsProductsLimitForRemoval(CategoryInterface $category, $includeSub)
+    {
+        return $this->countProducts($category, $includeSub, true) > $this->productsLimitForRemoval;
```
Comment (solivier) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: are you sure this works ? it seems to miss the "-" of the "->"

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php (hunk 1 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: one line

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php (hunk 2 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: update phpdoc

**Comment by nidup** [comment]
please double check if CategoryExtension is overriden in EE and if there is no impacts with permissions otherwise GTM

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php (hunk 4 above)
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: a comparison dude ;)

File: src/Pim/Bundle/EnrichBundle/Resources/views/CategoryTree/form.html.twig [hunk 5]
```diff
@@ -27,15 +27,31 @@
...
+                        'attr': {
+                            'data-title':       'confirmation.delete'|trans,
+                            'data-dialog':      'alert',
+                            'data-message':     'info.category.products limit exceeded'|trans({ '%limit%': get_products_limit_for_removal() })
+                        }
```
Comment (jmleroux) [inline src/Pim/Bundle/EnrichBundle/Resources/views/CategoryTree/form.html.twig]: too much spaces

File: src/Pim/Bundle/EnrichBundle/Resources/translations/messages.en.yml [hunk 6]
```diff
@@ -203,8 +203,9 @@ flash:
 info:
     updated: There are unsaved changes.
     category:
-        remove children: All sub-categories will be deleted.
-        keep products:   Products in these categories will not be deleted.
+        remove children:          All sub-categories will be deleted.
+        keep products:            Products in these categories will not be deleted.
+        products limit exceeded:  This category contains more products than allowed for this operation (%limit% products max.)
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Resources/translations/messages.en.yml]: maximum
not max.

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 7]
```diff
@@ -35,12 +41,14 @@ public function __construct(ProductCategoryManager $manager)
...
-            'children_tree_response'   => new \Twig_Function_Method($this, 'childrenTreeResponse'),
-            'list_categories_response' => new \Twig_Function_Method($this, 'listCategoriesResponse'),
-            'list_trees_response'      => new \Twig_Function_Method($this, 'listTreesResponse')
-        );
+        return [
+            'children_response'                             => new \Twig_Function_Method($this, 'childrenResponse'),
+            'children_tree_response'                        => new \Twig_Function_Method($this, 'childrenTreeResponse'),
+            'list_categories_response'                      => new \Twig_Function_Method($this, 'listCategoriesResponse'),
+            'list_trees_response'                           => new \Twig_Function_Method($this, 'listTreesResponse'),
+            'category_exceeds_products_limit_for_removal'   => new \Twig_Function_Method($this, 'categoryExceedsProductsLimitForRemoval'),
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: extra space

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php (hunk 1 above)
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: @var int

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 8]
```diff
@@ -21,13 +21,19 @@ class CategoryExtension extends \Twig_Extension
     protected $manager;
 
     /**
+     * @var integer
+     */
```
Comment (jmleroux) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: `/** @var integer */`

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 9]
```diff
@@ -162,6 +170,29 @@ public function listCategoriesResponse(
     }
 
     /**
+     * Check if specified category exceeds the products limit for removal
+     *
+     * @param CategoryInterface $category
+     * @param boolean           $includeSub
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: we use `@bool` ;)

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 10]
```diff
@@ -162,6 +170,29 @@ public function listCategoriesResponse(
...
+     *
+     * @return boolean
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: same

File: app/config/pim_parameters.yml [hunk 11]
```diff
@@ -24,8 +24,9 @@ parameters:
     # use PimInstallerBundle:minimal for minimal data set
     installer_data:       PimInstallerBundle:icecat_demo_dev
 
-    uservoice_key:        ~
-    mass_edit_limit:      1000
+    uservoice_key:                              ~
+    mass_edit_limit:                            1000
+    linked_products_limit_for_category_removal: 100
```
Comment (skeleton) [inline app/config/pim_parameters.yml]: no way to reduce this key ?

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php (hunk 10 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: @return bool

File: src/Pim/Bundle/VersioningBundle/UpdateGuesser/MongoDBODM/ContainsProductsUpdateGuesser.php [hunk 12]
```diff
@@ -51,8 +52,9 @@ public function guessUpdates(EntityManager $em, $entity, $action)
             foreach ($products as $product) {
                 $pendings[] = $product;
             }
-
-        } elseif ($entity instanceof CategoryInterface && $entity->getId()) {
+        } elseif ($entity instanceof CategoryInterface &&
+                  $entity->getId() &&
+                  UpdateGuesserInterface::ACTION_DELETE === $action) {
```
Comment (grena) [inline src/Pim/Bundle/VersioningBundle/UpdateGuesser/MongoDBODM/ContainsProductsUpdateGuesser.php]: please put the opening brace on a new line

```php
        } elseif ($entity instanceof CategoryInterface &&
                  $entity->getId() &&
                  UpdateGuesserInterface::ACTION_DELETE === $action
        ) {
```

File: src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php [hunk 13]
```diff
@@ -162,6 +170,29 @@ public function listCategoriesResponse(
...
+     */
+    public function categoryExceedsProductsLimitForRemoval(CategoryInterface $category, $includeSub)
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Twig/CategoryExtension.php]: no way to reduce the name of this method ?

File: app/config/pim_parameters.yml (hunk 11 above)
Comment (phaseinducer) [inline app/config/pim_parameters.yml]: I could remove "linked". If you have any idea...

File: src/Pim/Bundle/VersioningBundle/UpdateGuesser/MongoDBODM/ContainsProductsUpdateGuesser.php (hunk 12 above)
Comment (jjanvier) [inline src/Pim/Bundle/VersioningBundle/UpdateGuesser/MongoDBODM/ContainsProductsUpdateGuesser.php]: you can use php-cs-fixer or reformat the code with phpstorm

File: app/config/pim_parameters.yml (hunk 11 above)
Comment (skeleton) [inline app/config/pim_parameters.yml]: max_products_category_removal ?

Comment (phaseinducer) [inline app/config/pim_parameters.yml]: ok thx =)

File: spec/Pim/Bundle/EnrichBundle/Twig/CategoryExtensionSpec.php [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+
+namespace spec\Pim\Bundle\EnrichBundle\Twig;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use PhpSpec\ObjectBehavior;
+use Pim\Bundle\CatalogBundle\Entity\Category;
+use Pim\Bundle\CatalogBundle\Manager\ProductCategoryManager;
+use Prophecy\Argument;
+
+
```
Comment (skeleton) [inline spec/Pim/Bundle/EnrichBundle/Twig/CategoryExtensionSpec.php]: remove extra blank line

File: spec/Pim/Bundle/EnrichBundle/Twig/CategoryExtensionSpec.php [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+                        'data'                  => 'Some category 1',
+                        'state'                 => 'leaf',
+                        'children'              => [],
+                        'selectedChildrenCount' => 0
+                    ],
+                    [
+                        'attr'                  => ['id' => 'node_3'],
+                        'data'                  => 'Some category 2',
+                        'state'                 => 'leaf',
+                        'children'              =>[],
```
Comment (skeleton) [inline spec/Pim/Bundle/EnrichBundle/Twig/CategoryExtensionSpec.php]: add space after "=>"

**Comment by nidup** [comment]
GTM

**Pull request merged**